Crash because of storage.List

### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

We used the Fuzz engine to modify some Test(TestInMemoryRepositoryListing) data, and then the following crash appeared. We debugged and eventually found the crash point is at line 300 of `internal/repository/json.go`. We feel that this could be an out-of-boundary bug.

### How to reproduce

The simplified test code is in **Example code**

### Screenshots

Crash log
```
panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1]

goroutine 19 [running]:
testing.tRunner.func1.2({0x3f3d60, 0xc0000a2780})
	D:/.../go1.19/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	D:/.../go1.19/src/testing/testing.go:1399 +0x39f
panic({0x3f3d60, 0xc0000a2780})
	D:/.../go1.19/src/runtime/panic.go:884 +0x212
fyne.io/fyne/v2/internal/repository.(*InMemoryRepository).List(0xc00009c258, {0x4b9518?, 0xc0000e4690?})
	D:/.../fyne/internal/repository/memory.go:300 +0x32e
fyne.io/fyne/v2/storage.List({0x4b9518, 0xc0000e4690})
	D:/.../fyne/storage/uri.go:479 +0x58
fyne.io/fyne/v2/internal/repository.TestInMemoryRepositoryListing2(0x0?)
	D:/.../fyne/internal/repository/InMemoryRepositoryListing_fuzz_test.go:41 +0xfc
testing.tRunner(0xc000084d00, 0x44c160)
	D:/.../go1.19/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	D:/.../go1.19/src/testing/testing.go:1493 +0x35f
```

### Example code

```go
m := NewInMemoryRepository("mem")
repository.Register("mem", m)
m.Data[""] = []byte{1, 2, 3}
foo, _ := storage.ParseURI("mem:")
storage.List(foo)
```

### Fyne version

commit [cfb88d7](https://github.com/fyne-io/fyne/commit/cfb88d7c4038f737f94ca702f7cf4e5d6c86e01d)

### Go compiler version

1.19

### Operating system and version

Windows 10

### Additional Information

_No response_

--- a/internal/repository/memory.go
+++ b/internal/repository/memory.go
@@ -1,13 +1,13 @@
 package repository
 
 import (
-	"fyne.io/fyne/v2"
-	"fyne.io/fyne/v2/storage"
-	"fyne.io/fyne/v2/storage/repository"
-
 	"fmt"
 	"io"
 	"strings"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/storage"
+	"fyne.io/fyne/v2/storage/repository"
 )
 
 // declare conformance to interfaces
@@ -297,7 +297,7 @@ func (m *InMemoryRepository) List(u fyne.URI) ([]fyne.URI, error) {
 		// does not have one.
 		pSplit := strings.Split(p, "/")
 		ncomp := len(pSplit)
-		if p[len(p)-1] == '/' {
+		if len(p) > 0 && p[len(p)-1] == '/' {
 			ncomp--
 		}
 

--- a/internal/repository/memory_test.go
+++ b/internal/repository/memory_test.go
@@ -68,6 +68,10 @@ func TestInMemoryRepositoryParsing(t *testing.T) {
 	baz, _ := storage.ParseURI("mem:///baz")
 	assert.Nil(t, err)
 	assert.NotNil(t, baz)
+
+	empty, _ := storage.ParseURI("mem:")
+	assert.Nil(t, err)
+	assert.NotNil(t, empty)
 }
 
 func TestInMemoryRepositoryExists(t *testing.T) {
@@ -328,6 +332,7 @@ func TestInMemoryRepositoryListing(t *testing.T) {
 	// set up our repository - it's OK if we already registered it
 	m := NewInMemoryRepository("mem")
 	repository.Register("mem", m)
+	m.Data[""] = []byte{1, 2, 3}
 	m.Data["/foo"] = []byte{1, 2, 3}
 	m.Data["/foo/bar"] = []byte{1, 2, 3}
 	m.Data["/foo/baz/"] = []byte{1, 2, 3}
@@ -346,6 +351,11 @@ func TestInMemoryRepositoryListing(t *testing.T) {
 		stringListing = append(stringListing, u.String())
 	}
 	assert.ElementsMatch(t, []string{"mem:///foo/bar", "mem:///foo/baz/"}, stringListing)
+
+	empty, _ := storage.ParseURI("mem:") // invalid path
+	canList, err = storage.CanList(empty)
+	assert.NotNil(t, err)
+	assert.False(t, canList)
 }
 
 func TestInMemoryRepositoryCreateListable(t *testing.T) {

